cache-utils.go

```diff
@@ -0,0 +1,140 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+)
+
+// ResponseRecorder returns a wrapped response writer to
+// get underlying http.Response for cache handler.
+type ResponseRecorder struct {
+	StatusCode   int
+	header       http.Header // header needed for Header() method
+	Flushed      bool        // Flushed is whether the Handler called Flush.
+	ch           chan bool   // blocks until header is written or Write started
+	sentResponse bool
+	wroteHeader  bool
+
+	pw     *io.PipeWriter
+	result *http.Response // http response
+}
+
+// NewRecorder returns an initialized ResponseRecorder.
+func NewRecorder() *ResponseRecorder {
+	pr, pw := io.Pipe()
+	nr := &ResponseRecorder{
+		ch:         make(chan bool),
+		header:     make(http.Header),
+		StatusCode: 200,
+		pw:         pw,
+		result: &http.Response{
+			Proto:      "HTTP/1.1",
+			ProtoMajor: 1,
+			ProtoMinor: 1,
+			Header:     make(http.Header),
+			Close:      true,
+			Body:       pr,
+		},
+	}
+	return nr
+}
+
+// Header needed for implementing "net/http".ResponseWriter
+func (rw *ResponseRecorder) Header() http.Header {
+	return rw.header
+}
+
+// writeHeader writes status code if no header written yet.
+func (rw *ResponseRecorder) writeHeader(b []byte) {
+	if rw.wroteHeader {
+		return
+	}
+	rw.WriteHeader(200)
+}
+
+// Write implements http.ResponseWriter. The data in buf is written to
+// the pipeWriter
+func (rw *ResponseRecorder) Write(buf []byte) (int, error) {
+	rw.writeHeader(buf)
+	if !rw.sentResponse {
+		rw.sendResponse()
+	}
+	return rw.pw.Write(buf)
+}
+
+func (rw *ResponseRecorder) sendResponse() {
+	if rw.sentResponse {
+		return
+	}
+	rw.sentResponse = true
+	rw.ch <- true
+}
+
+// cleanup pipe writer after proxy ServeHTTP finishes.
+func (rw *ResponseRecorder) finish() {
+	if !rw.wroteHeader {
+		rw.WriteHeader(200)
+	}
+	if !rw.sentResponse {
+		rw.sendResponse()
+	}
+	rw.pw.Close()
+	rw.Flush()
+}
+
+// WriteHeader implements http.ResponseWriter.
+func (rw *ResponseRecorder) WriteHeader(code int) {
+	if rw.wroteHeader {
+		return
+	}
+	rw.StatusCode = code
+	rw.wroteHeader = true
+	if rw.header == nil {
+		rw.header = make(http.Header)
+	}
+	if !rw.sentResponse {
+		rw.sendResponse()
+	}
+}
+
+// Flush calls underlying Flush method
+func (rw *ResponseRecorder) Flush() {
+	if !rw.wroteHeader {
+		rw.WriteHeader(200)
+	}
+	rw.Flushed = true
+}
+
+// Result returns the response generated by the handler. It blocks on the
+// rw.ch until header some content has been written
+// The returned Response will have at least its StatusCode,
+// Header, Body
+func (rw *ResponseRecorder) Result() *http.Response {
+	// block until header/writes are started
+	<-rw.ch
+
+	res := &http.Response{
+		Proto:      "HTTP/1.1",
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		StatusCode: rw.StatusCode,
+		Header:     rw.header,
```

Comment:
This should be a clone `rw.header.Clone()`

Reply:
fixed.